fabric-registry-sync-v0/src/main/java/net/fabricmc/fabric/impl/registry/sync/RegistrySyncManager.java

```diff
@@ -38,55 +38,70 @@
 import org.jetbrains.annotations.Nullable;
 
 import net.minecraft.nbt.NbtCompound;
-import net.minecraft.network.Packet;
-import net.minecraft.util.Identifier;
 import net.minecraft.network.PacketByteBuf;
+import net.minecraft.server.network.ServerPlayerEntity;
+import net.minecraft.util.Identifier;
 import net.minecraft.util.registry.Registry;
 import net.minecraft.util.thread.ThreadExecutor;
 
 import net.fabricmc.fabric.api.event.registry.RegistryAttribute;
 import net.fabricmc.fabric.api.event.registry.RegistryAttributeHolder;
 import net.fabricmc.fabric.api.networking.v1.ServerPlayNetworking;
+import net.fabricmc.fabric.impl.registry.sync.map.IdMap;
+import net.fabricmc.fabric.impl.registry.sync.map.RegistryMap;
+import net.fabricmc.fabric.impl.registry.sync.packet.DirectRegistrySyncPacket;
+import net.fabricmc.fabric.impl.registry.sync.packet.NbtRegistrySyncPacket;
+import net.fabricmc.fabric.impl.registry.sync.packet.RegistrySyncPacket;
 
 public final class RegistrySyncManager {
 	static final boolean DEBUG = System.getProperty("fabric.registry.debug", "false").equalsIgnoreCase("true");
-	static final Identifier ID = new Identifier("fabric", "registry/sync");
 	private static final Logger LOGGER = LogManager.getLogger("FabricRegistrySync");
 	private static final boolean DEBUG_WRITE_REGISTRY_DATA = System.getProperty("fabric.registry.debug.writeContentsAsCsv", "false").equalsIgnoreCase("true");
+	private static final boolean FORCE_NBT_SYNC = System.getProperty("fabric.registry.forceNbtSync", "false").equalsIgnoreCase("true");
 
 	//Set to true after vanilla's bootstrap has completed
 	public static boolean postBootstrap = false;
 
 	private RegistrySyncManager() { }
 
-	public static Packet<?> createPacket() {
-		LOGGER.debug("Creating registry sync packet");
-
-		NbtCompound tag = toTag(true, null);
+	public static void sendPacket(ServerPlayerEntity player) {
+		if (FORCE_NBT_SYNC) {
+			LOGGER.warn("Force NBT sync is enabled");
+			sendPacket(player, NbtRegistrySyncPacket.getInstance());
+			return;
+		}
 
-		if (tag == null) {
-			return null;
+		if (ServerPlayNetworking.canSend(player, DirectRegistrySyncPacket.ID)) {
+			sendPacket(player, DirectRegistrySyncPacket.getInstance());
+		} else {
+			LOGGER.warn("Player {} can't receive direct packet, using nbt packet instead", player.getEntityName());
+			sendPacket(player, NbtRegistrySyncPacket.getInstance());
```

Comment:
Should we consider kicking the client if it cannot accept the legacy registry sync as well, as otherwise the client will have a glitchy experience?